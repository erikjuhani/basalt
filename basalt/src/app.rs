use crate::markdown::{MarkdownView, MarkdownViewState};
use basalt_core::obsidian::{Note, Vault};
use crossterm::event::{self, Event, KeyEvent, KeyEventKind};
use ratatui::{
    buffer::Buffer,
    layout::{Constraint, Layout, Rect, Size},
    widgets::{StatefulWidget, StatefulWidgetRef},
    DefaultTerminal,
};

use std::{cell::RefCell, io::Result, marker::PhantomData};

use crate::{
    config::{Config, Key},
    help_modal::{HelpModal, HelpModalState},
    sidepanel::{SidePanel, SidePanelState},
    start::{StartScreen, StartState},
    statusbar::{StatusBar, StatusBarState},
    text_counts::{CharCount, WordCount},
    vault_selector_modal::{VaultSelectorModal, VaultSelectorModalState},
};

const VERSION: &str = env!("CARGO_PKG_VERSION");

const HELP_TEXT: &str = include_str!("./help.txt");

#[derive(Debug, Clone, Default, PartialEq)]
pub enum Mode {
    #[default]
    Select,
    Normal,
    Insert,
}

impl Mode {
    fn as_str(&self) -> &'static str {
        match self {
            Mode::Select => "Select",
            Mode::Normal => "Normal",
            Mode::Insert => "Insert",
        }
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq)]
pub enum ScrollAmount {
    #[default]
    One,
    HalfPage,
}

fn calc_scroll_amount(scroll_amount: ScrollAmount, size: Size) -> usize {
    match scroll_amount {
        ScrollAmount::One => 1,
        ScrollAmount::HalfPage => (size.height / 3).into(),
    }
}

#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Action {
    Select,
    Next,
    Prev,
    Insert,
    Resize(Size),
    ScrollUp(ScrollAmount),
    ScrollDown(ScrollAmount),
    ToggleMode,
    ToggleHelp,
    ToggleVaultSelector,
    Quit,
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct Start<'a> {
    pub start_state: StartState<'a>,
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct Main<'a> {
    pub sidepanel_state: SidePanelState<'a>,
    pub selected_note: Option<SelectedNote>,
    pub markdown_view_state: MarkdownViewState,
    pub notes: Vec<Note>,
    pub vaults: Vec<&'a Vault>,
    pub size: Size,
    pub mode: Mode,
}

impl<'a> Main<'a> {
    fn new(vault_name: &'a str, notes: Vec<Note>, size: Size, vaults: Vec<&'a Vault>) -> Self {
        Self {
            notes: notes.clone(),
            sidepanel_state: SidePanelState::new(vault_name, notes),
            vaults,
            size,
            ..Default::default()
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub enum Screen<'a> {
    Start(Start<'a>),
    Main(Main<'a>),
}

impl Default for Screen<'_> {
    fn default() -> Self {
        Screen::Start(Start::default())
    }
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct AppState<'a> {
    pub config: Config,
    pub help_modal: Option<HelpModalState>,
    pub vault_selector_modal: Option<VaultSelectorModalState<'a>>,
    pub size: Size,
    pub is_running: bool,
    pub screen: Screen<'a>,
    _lifetime: PhantomData<&'a ()>,
}

pub struct App<'a> {
    pub state: AppState<'a>,
    terminal: RefCell<DefaultTerminal>,
}

impl<'a> StatefulWidgetRef for App<'a> {
    type State = AppState<'a>;

    fn render_ref(&self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let screen = state.screen.clone();

        match screen {
            Screen::Start(mut state) => {
                StartScreen::default().render_ref(area, buf, &mut state.start_state)
            }
            Screen::Main(mut state) => {
                let [content, statusbar] =
                    Layout::vertical([Constraint::Fill(1), Constraint::Length(1)])
                        .horizontal_margin(1)
                        .areas(area);

                let (left, right) = if state.mode == Mode::Select {
                    (Constraint::Length(35), Constraint::Fill(1))
                } else {
                    (Constraint::Length(5), Constraint::Fill(1))
                };

                let [sidepanel, note] = Layout::horizontal([left, right]).areas(content);

                SidePanel::default().render_ref(sidepanel, buf, &mut state.sidepanel_state);

                MarkdownView.render_ref(note, buf, &mut state.markdown_view_state);

                let mode = state.mode.as_str().to_uppercase();
                let (name, counts) = state
                    .selected_note
                    .clone()
                    .map(|note| {
                        let content = note.content.as_str();
                        (
                            note.name,
                            (WordCount::from(content), CharCount::from(content)),
                        )
                    })
                    .unzip();

                let (word_count, char_count) = counts.unwrap_or_default();

                let mut status_bar_state = StatusBarState::new(
                    &mode,
                    name.as_deref(),
                    word_count.into(),
                    char_count.into(),
                );

                StatusBar::default().render_ref(statusbar, buf, &mut status_bar_state);
            }
        }

        if let Some(mut vault_selector_modal_state) = state.vault_selector_modal.clone() {
            VaultSelectorModal::default().render(area, buf, &mut vault_selector_modal_state)
        }

        if let Some(mut help_modal_state) = state.help_modal.clone() {
            HelpModal.render(area, buf, &mut help_modal_state)
        }
    }
}

#[derive(Debug, Default, Clone, PartialEq)]
pub struct SelectedNote {
    name: String,
    path: String,
    content: String,
}

impl From<&Note> for SelectedNote {
    fn from(value: &Note) -> Self {
        Self {
            name: value.name.clone(),
            path: value.path.to_string_lossy().to_string(),
            content: Note::read_to_string(value).unwrap(),
        }
    }
}

fn help_text() -> String {
    let version = format!("{VERSION}~alpha");
    HELP_TEXT.replace(
        "%version-notice",
        format!("This is the read-only release of Basalt ({version})").as_str(),
    )
}

impl<'a> App<'a> {
    pub fn start(terminal: DefaultTerminal, vaults: Vec<&Vault>) -> Result<()> {
        let version = format!("{VERSION}~alpha");
        let size = terminal.size()?;

        let state = AppState {
            config: Config::build(),
            screen: Screen::Start(Start {
                start_state: StartState::new(&version, size, vaults),
            }),
            size,
            is_running: true,
            _lifetime: PhantomData,
            ..Default::default()
        };

        App {
            state: state.clone(),
            terminal: RefCell::new(terminal),
        }
        .run(state)
    }

    fn run(&mut self, state: AppState<'a>) -> Result<()> {
        self.draw(&state)?;
        let event = &event::read()?;
        match self.update(&state, self.handle_event(event)) {
            state if state.is_running => self.run(state),
            _ => Ok(()),
        }
    }

    fn update_help_modal(
        &self,
        state: AppState<'a>,
        inner: HelpModalState,
        action: Option<Action>,
    ) -> AppState<'a> {
        match action {
            Some(Action::ScrollUp(amount)) => AppState {
                help_modal: Some(inner.scroll_up(calc_scroll_amount(amount, state.size))),
                ..state
            },
            Some(Action::ScrollDown(amount)) => AppState {
                help_modal: Some(inner.scroll_down(calc_scroll_amount(amount, state.size))),
                ..state
            },
            Some(Action::Next) => AppState {
                help_modal: Some(inner.scroll_down(1)),
                ..state
            },
            Some(Action::Prev) => AppState {
                help_modal: Some(inner.scroll_up(1)),
                ..state
            },
            _ => state,
        }
    }

    fn update_vault_selector_modal(
        &self,
        state: AppState<'a>,
        inner: VaultSelectorModalState<'a>,
        action: Option<Action>,
    ) -> AppState<'a> {
        match action {
            Some(Action::ToggleVaultSelector) => AppState {
                vault_selector_modal: None,
                ..state
            },
            Some(Action::Select) => {
                // TODO: Add logic to not load the vault again if the same vault was picked in the
                // selector.
                let alphabetically =
                    |a: &Note, b: &Note| a.name.to_lowercase().cmp(&b.name.to_lowercase());

                let vault_selector_state = inner.vault_selector_state.select();

                let vault_with_notes = vault_selector_state
                    .selected()
                    .and_then(|index| inner.vault_selector_state.get_item(index))
                    .map(|vault| (vault, vault.notes_sorted_by(alphabetically)));

                if let Some((vault, notes)) = vault_with_notes {
                    AppState {
                        screen: Screen::Main(Main::new(
                            &vault.name,
                            notes,
                            state.size,
                            vault_selector_state.items(),
                        )),
                        vault_selector_modal: None,
                        ..state
                    }
                } else {
                    state
                }
            }
            Some(Action::Next) => AppState {
                vault_selector_modal: Some(VaultSelectorModalState {
                    vault_selector_state: inner.vault_selector_state.next(),
                }),
                ..state
            },
            Some(Action::Prev) => AppState {
                vault_selector_modal: Some(VaultSelectorModalState {
                    vault_selector_state: inner.vault_selector_state.previous(),
                }),
                ..state
            },
            _ => state,
        }
    }

    fn update_select_mode(
        &self,
        state: AppState<'a>,
        inner: Main<'a>,
        action: Option<Action>,
    ) -> AppState<'a> {
        match action {
            Some(Action::ToggleMode) => AppState {
                screen: Screen::Main(Main {
                    mode: Mode::Normal,
                    sidepanel_state: inner.sidepanel_state.close(),
                    ..inner
                }),
                ..state
            },
            Some(Action::ScrollUp(amount)) => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner
                        .markdown_view_state
                        .scroll_up(calc_scroll_amount(amount, state.size)),
                    ..inner
                }),
                ..state
            },
            Some(Action::ScrollDown(amount)) => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner
                        .markdown_view_state
                        .scroll_down(calc_scroll_amount(amount, state.size)),
                    ..inner
                }),
                ..state
            },
            Some(Action::Select) => {
                let sidepanel_state = inner.sidepanel_state.select();

                let selected_note = inner
                    .notes
                    .get(sidepanel_state.selected().unwrap_or_default())
                    .map(SelectedNote::from);

                AppState {
                    screen: Screen::Main(Main {
                        sidepanel_state,
                        selected_note: selected_note.clone(),
                        markdown_view_state: inner
                            .markdown_view_state
                            .set_text(selected_note.map(|note| note.content).unwrap_or_default())
                            .reset_scrollbar(),
                        ..inner
                    }),
                    ..state
                }
            }
            Some(Action::Next) => AppState {
                screen: Screen::Main(Main {
                    sidepanel_state: inner.sidepanel_state.next(),
                    ..inner
                }),
                ..state
            },
            Some(Action::Prev) => AppState {
                screen: Screen::Main(Main {
                    sidepanel_state: inner.sidepanel_state.previous(),
                    ..inner
                }),
                ..state
            },
            _ => state,
        }
    }

    fn update_normal_mode(
        &self,
        state: AppState<'a>,
        inner: Main<'a>,
        action: Option<Action>,
    ) -> AppState<'a> {
        let Some(action) = action else {
            return state;
        };

        match action {
            Action::ToggleMode => AppState {
                screen: Screen::Main(Main {
                    mode: Mode::Select,
                    sidepanel_state: inner.sidepanel_state.open(),
                    ..inner
                }),
                ..state
            },
            Action::ScrollUp(amount) => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner
                        .markdown_view_state
                        .scroll_up(calc_scroll_amount(amount, state.size)),
                    ..inner
                }),
                ..state
            },
            Action::ScrollDown(amount) => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner
                        .markdown_view_state
                        .scroll_down(calc_scroll_amount(amount, state.size)),
                    ..inner
                }),
                ..state
            },
            Action::Next => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner.markdown_view_state.scroll_down(1),
                    ..inner
                }),
                ..state
            },
            Action::Prev => AppState {
                screen: Screen::Main(Main {
                    markdown_view_state: inner.markdown_view_state.scroll_up(1),
                    ..inner
                }),
                ..state
            },
            _ => state,
        }
    }

    fn update_main_state(
        &self,
        state: AppState<'a>,
        inner: Main<'a>,
        action: Option<Action>,
    ) -> AppState<'a> {
        if let Some(Action::ToggleVaultSelector) = action {
            return AppState {
                vault_selector_modal: if state.vault_selector_modal.is_some() {
                    None
                } else {
                    Some(VaultSelectorModalState::new(inner.vaults.clone()))
                },
                ..state
            };
        }

        match inner.mode {
            Mode::Select => self.update_select_mode(state, inner, action),
            Mode::Normal => self.update_normal_mode(state, inner, action),
            Mode::Insert => state,
        }
    }

    fn update_start_state(
        &self,
        state: AppState<'a>,
        inner: Start<'a>,
        action: Option<Action>,
    ) -> AppState<'a> {
        match action {
            Some(Action::Select) => {
                let alphabetically =
                    |a: &Note, b: &Note| a.name.to_lowercase().cmp(&b.name.to_lowercase());

                let splash_state = inner.start_state.select();

                let vault_with_notes = splash_state
                    .selected()
                    .and_then(|index| splash_state.get_item(index))
                    .map(|vault| (vault, vault.notes_sorted_by(alphabetically)));

                if let Some((vault, notes)) = vault_with_notes {
                    AppState {
                        screen: Screen::Main(Main::new(
                            &vault.name,
                            notes,
                            state.size,
                            inner.start_state.items(),
                        )),
                        ..state
                    }
                } else {
                    state
                }
            }
            Some(Action::Next) => AppState {
                screen: Screen::Start(Start {
                    start_state: inner.start_state.next(),
                }),
                ..state
            },
            Some(Action::Prev) => AppState {
                screen: Screen::Start(Start {
                    start_state: inner.start_state.previous(),
                }),
                ..state
            },
            _ => state,
        }
    }

    fn update(&self, state: &AppState<'a>, action: Option<Action>) -> AppState<'a> {
        let state = state.clone();
        let screen = state.screen.clone();
        match action {
            Some(Action::Quit) => AppState {
                is_running: false,
                ..state
            },
            Some(Action::ToggleHelp) => AppState {
                help_modal: if state.help_modal.is_some() {
                    None
                } else {
                    Some(HelpModalState::new(&help_text()))
                },
                ..state
            },
            Some(Action::Resize(size)) => AppState { size, ..state },
            _ if state.help_modal.is_some() => {
                self.update_help_modal(state.clone(), state.help_modal.unwrap().clone(), action)
            }
            _ if state.vault_selector_modal.is_some() => self.update_vault_selector_modal(
                state.clone(),
                state.vault_selector_modal.unwrap().clone(),
                action,
            ),
            _ => match screen {
                Screen::Start(inner) => self.update_start_state(state, inner, action),
                Screen::Main(inner) => self.update_main_state(state, inner, action),
            },
        }
    }

    fn handle_event(&self, event: &Event) -> Option<Action> {
        match event {
            Event::Resize(cols, rows) => Some(Action::Resize(Size::new(*cols, *rows))),
            Event::Key(key_event) if key_event.kind == KeyEventKind::Press => {
                self.handle_press_key_event(key_event)
            }
            _ => None,
        }
    }

    fn handle_press_key_event(&self, key_event: &KeyEvent) -> Option<Action> {
        let key_binding: Key = key_event.into();

        self.state
            .config
            .keymap
            .get(&key_binding)
            .cloned()
            .or_else(|| key_binding.eq(&Key::CTRLC).then_some(Action::Quit))
    }

    fn draw(&self, state: &AppState<'a>) -> Result<()> {
        let mut terminal = self.terminal.borrow_mut();
        let mut state = state.clone();

        terminal.draw(move |frame| {
            let area = frame.area();
            let buf = frame.buffer_mut();
            self.render_ref(area, buf, &mut state);
        })?;

        Ok(())
    }
}
