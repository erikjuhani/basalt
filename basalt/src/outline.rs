use item::{Flatten, Item};
pub use state::OutlineState;

mod item;
mod state;

use ratatui::{
    buffer::Buffer,
    layout::{Alignment, Rect},
    style::{Style, Stylize},
    text::{Line, Span},
    widgets::{Block, BorderType, Borders, List, ListItem, Padding, StatefulWidget},
};

use crate::{
    app::{ActivePane, Message as AppMessage},
    explorer,
    note_editor::{self, markdown_parser::Node},
};

#[derive(Clone, Debug, PartialEq)]
pub enum Message {
    Up,
    Down,
    Select,
    SelectAt(usize),
    SetNodes(Vec<Node>),
    Expand,
    Toggle,
    ToggleExplorer,
    SwitchPaneNext,
    SwitchPanePrevious,
}

pub fn update<'a>(message: &Message, state: &mut OutlineState) -> Option<AppMessage<'a>> {
    match message {
        Message::Up => state.previous(1),
        Message::Down => state.next(1),
        Message::Expand => state.toggle_item(),
        Message::SelectAt(index) => state.select_at(*index),
        Message::SetNodes(nodes) => state.set_nodes(nodes),

        Message::SwitchPaneNext => {
            state.set_active(false);
            return Some(AppMessage::SetActivePane(ActivePane::Explorer));
        }
        Message::SwitchPanePrevious => {
            state.set_active(false);
            return Some(AppMessage::SetActivePane(ActivePane::NoteEditor));
        }
        Message::Toggle => {
            state.toggle();
            if !state.is_open() {
                state.set_active(false);
                return Some(AppMessage::SetActivePane(ActivePane::NoteEditor));
            }
        }
        Message::Select => {
            if let Some(item) = state.selected() {
                return Some(AppMessage::NoteEditor(note_editor::Message::SetRow(
                    item.get_range().start,
                )));
            }
        }
        Message::ToggleExplorer => {
            return Some(AppMessage::Explorer(explorer::Message::Toggle));
        }
    };

    None
}

#[derive(Default)]
pub struct Outline;

trait AsListItems {
    fn to_list_items(&self) -> Vec<ListItem<'_>>;
    fn to_collapsed_items(&self) -> Vec<ListItem<'_>>;
}

impl AsListItems for Vec<Item> {
    fn to_collapsed_items(&self) -> Vec<ListItem<'_>> {
        self.flatten()
            .iter()
            .map(|item| match item {
                Item::Heading { .. } => ListItem::new(Line::from("·")).dark_gray().dim(),
                Item::HeadingEntry { expanded: true, .. } => {
                    ListItem::new(Line::from("✺")).red().dim()
                }
                Item::HeadingEntry {
                    expanded: false, ..
                } => ListItem::new(Line::from("◦")).dark_gray().dim(),
            })
            .collect()
    }

    fn to_list_items(&self) -> Vec<ListItem<'_>> {
        fn list_item<'a>(indentation: Span<'a>, symbol: &'a str, content: &'a str) -> ListItem<'a> {
            ListItem::new(Line::from(
                [indentation, symbol.into(), content.into()].to_vec(),
            ))
        }

        fn to_list_items(depth: usize) -> impl Fn(&Item) -> Vec<ListItem> {
            let indentation = if depth > 0 {
                Span::raw("│ ".repeat(depth)).black()
            } else {
                Span::raw("  ".repeat(depth)).black()
            };
            move |item| match item {
                Item::Heading { content, .. } => {
                    vec![list_item(indentation.clone(), "  ", content)]
                }
                Item::HeadingEntry {
                    expanded: true,
                    children,
                    content,
                    ..
                } => {
                    let mut items = vec![list_item(indentation.clone(), "▾ ", content)];
                    items.extend(children.iter().flat_map(to_list_items(depth + 1)));
                    items
                }
                Item::HeadingEntry {
                    expanded: false,
                    content,
                    ..
                } => vec![list_item(indentation.clone(), "▸ ", content)],
            }
        }

        self.iter().flat_map(to_list_items(0)).collect()
    }
}

impl StatefulWidget for Outline {
    type State = OutlineState;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let block = Block::bordered()
            .border_type(if state.active {
                BorderType::Thick
            } else {
                BorderType::Rounded
            })
            .title(if state.is_open() {
                " ▶ Outline "
            } else {
                " ◀ "
            })
            .title_alignment(Alignment::Right)
            .padding(Padding::horizontal(1))
            .title_style(Style::default().italic().bold());

        let items = if state.is_open() {
            state.items.to_list_items()
        } else {
            state.items.to_collapsed_items()
        };

        List::new(items)
            .block(if state.is_open() {
                block
            } else {
                block.borders(Borders::RIGHT | Borders::TOP | Borders::BOTTOM)
            })
            .highlight_style(Style::default().reversed().dark_gray())
            .highlight_symbol("")
            .render(area, buf, &mut state.list_state);
    }
}

#[cfg(test)]
mod tests {
    use crate::note_editor::markdown_parser;

    use super::*;
    use indoc::indoc;
    use insta::assert_snapshot;
    use ratatui::{backend::TestBackend, Terminal};

    #[test]
    fn test_outline_render() {
        let tests = [
            ("empty", markdown_parser::from_str("")),
            ("single_level", markdown_parser::from_str("# Heading 1")),
            (
                "only_top_level",
                markdown_parser::from_str(indoc! {r#"
                # Heading 1
                # Heading 2
                # Heading 3
                # Heading 4
                # Heading 5
                # Heading 6
            "#}),
            ),
            (
                "only_deep_level",
                markdown_parser::from_str(indoc! {r#"
                ###### Heading 1
                ##### Heading 2
                ###### Heading 2.1
                ###### Heading 2.2
                ##### Heading 3
                ##### Heading 4
                ###### Heading 4.1
                ##### Heading 5
            "#}),
            ),
            (
                "sequential_all_levels",
                markdown_parser::from_str(indoc! {r#"
                # Heading 1
                ## Heading 2
                ### Heading 3
                #### Heading 4
                ##### Heading 5
                ###### Heading 6
            "#}),
            ),
            (
                "complex_nested_structure",
                markdown_parser::from_str(indoc! {r#"
                ## Heading 1
                ## Heading 2
                ### Heading 2.1
                #### Heading 2.1.1
                ### Heading 2.2
                #### Heading 2.2.1
                ## Heading 3
                ###### Heading 3.1.1.1.1.1
            "#}),
            ),
            (
                "irregular_nesting_with_skips",
                markdown_parser::from_str(indoc! {r#"
                # Heading 1
                ## Heading 2
                ## Heading 2.1
                #### Heading 2.1.1
                #### Heading 2.1.2
                ## Heading 2.2
                ### Heading 3
            "#}),
            ),
            (
                "level_skipping",
                markdown_parser::from_str(indoc! {r#"
                # Level 1
                ### Level 3 (skipped 2)
                ##### Level 5 (skipped 4)
                ## Level 2 (back to 2)
                ###### Level 6 (jump to 6)
            "#}),
            ),
            (
                "reverse_hierarchy",
                markdown_parser::from_str(indoc! {r#"
                ###### Level 6
                ##### Level 5
                #### Level 4
                ### Level 3
                ## Level 2
                # Level 1
            "#}),
            ),
            (
                "multiple_root_levels",
                markdown_parser::from_str(indoc! {r#"
                # Root 1
                ## Child 1.1
                ### Child 1.1.1

                ## Root 2 (different level)
                #### Child 2.1 (skipped level 3)

                ### Root 3 (different level)
                ###### Child 3.1 (deep skip)
            "#}),
            ),
            (
                "duplicate_headings",
                markdown_parser::from_str(indoc! {r#"
                # Duplicate
                ## Child
                # Duplicate
                ## Different Child
                # Duplicate
            "#}),
            ),
            (
                "mixed_with_content",
                markdown_parser::from_str(indoc! {r#"
                # Chapter 1
                Some paragraph content here.

                ## Section 1.1
                More content.

                - List item
                - Another item

                ### Subsection 1.1.1
                Final content.
            "#}),
            ),
            (
                "boundary_conditions_systematic",
                markdown_parser::from_str(indoc! {r#"
                # A
                ## B
                ### C
                #### D
                ##### E
                ###### F
                ##### E2
                #### D2
                ### C2
                ## B2
                # A2
            "#}),
            ),
        ];

        let mut terminal = Terminal::new(TestBackend::new(30, 10)).unwrap();

        tests.into_iter().for_each(|(name, nodes)| {
            _ = terminal.clear();
            let mut state = OutlineState::new(&nodes, 0, true);
            state.expand_all();
            terminal
                .draw(|frame| Outline.render(frame.area(), frame.buffer_mut(), &mut state))
                .unwrap();
            assert_snapshot!(name, terminal.backend());
        });
    }
}
