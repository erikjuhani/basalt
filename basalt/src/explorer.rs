mod item;
mod state;

pub use item::Item;
use ratatui::layout::Size;
use ratatui::widgets::Borders;
pub use state::ExplorerState;
pub use state::Sort;
pub use state::Visibility;

use std::{marker::PhantomData, path::PathBuf};

use basalt_core::obsidian::Note;
use ratatui::{
    buffer::Buffer,
    layout::{Alignment, Constraint, Layout, Rect},
    style::{Style, Stylize},
    text::{Line, Span},
    widgets::{Block, BorderType, List, ListItem, StatefulWidget},
};

use crate::app::{
    calc_scroll_amount, ActivePane, Message as AppMessage, ScrollAmount, SelectedNote,
};
use crate::outline;

const SORT_SYMBOL_ASC: &str = "‚ÜëùåÜ";
const SORT_SYMBOL_DESC: &str = "‚ÜìùåÜ";

#[derive(Clone, Debug, PartialEq)]
pub enum Message {
    Up,
    Down,
    Open,
    Sort,
    Toggle,
    ToggleOutline,
    HidePane,
    ExpandPane,
    SwitchPaneNext,
    SwitchPanePrevious,
    ScrollUp(ScrollAmount),
    ScrollDown(ScrollAmount),
}

pub fn update<'a>(
    message: &Message,
    screen_size: Size,
    state: &mut ExplorerState,
) -> Option<AppMessage<'a>> {
    match message {
        Message::Up => state.previous(1),
        Message::Down => state.next(1),
        Message::Sort => state.sort(),
        Message::Toggle => state.toggle(),
        Message::HidePane => state.hide_pane(),
        Message::ExpandPane => state.expand_pane(),
        Message::SwitchPaneNext => {
            state.set_active(false);
            return Some(AppMessage::SetActivePane(ActivePane::NoteEditor));
        }
        Message::SwitchPanePrevious => {
            state.set_active(false);
            return Some(AppMessage::SetActivePane(ActivePane::Outline));
        }
        Message::ScrollUp(scroll_amount) => {
            state.previous(calc_scroll_amount(scroll_amount, screen_size.height.into()));
        }
        Message::ScrollDown(scroll_amount) => {
            state.next(calc_scroll_amount(scroll_amount, screen_size.height.into()));
        }
        Message::ToggleOutline => {
            return Some(AppMessage::Outline(outline::Message::Toggle));
        }
        Message::Open => {
            state.select();
            let note = state.selected_note.as_ref()?;
            return Some(AppMessage::SelectNote(SelectedNote::from(note)));
        }
    };

    None
}

#[derive(Default)]
pub struct Explorer<'a> {
    _lifetime: PhantomData<&'a ()>,
}

impl Explorer<'_> {
    pub fn new() -> Self {
        Self {
            _lifetime: PhantomData::<&()>,
        }
    }

    fn list_item<'a>(
        selected_path: Option<PathBuf>,
        is_open: bool,
    ) -> impl Fn(&'a (Item, usize)) -> ListItem<'a> {
        move |(item, depth)| {
            let indentation = if *depth > 0 {
                Span::raw("‚îÇ ".repeat(*depth)).black()
            } else {
                Span::raw("  ".repeat(*depth)).black()
            };
            match item {
                Item::File(Note { path, name }) => {
                    let is_selected = selected_path
                        .as_ref()
                        .is_some_and(|selected| selected == path);
                    ListItem::new(Line::from(match (is_open, is_selected) {
                        (true, true) => [indentation, "‚óÜ ".into(), name.into()].to_vec(),
                        (true, false) => [indentation, "  ".into(), name.into()].to_vec(),
                        (false, true) => ["‚óÜ".into()].to_vec(),
                        (false, false) => ["‚ó¶".dark_gray()].to_vec(),
                    }))
                }
                Item::Directory { expanded, name, .. } => {
                    ListItem::new(Line::from(match (is_open, expanded) {
                        (true, true) => [indentation, "‚ñæ ".dark_gray(), name.into()].to_vec(),
                        (true, false) => [indentation, "‚ñ∏ ".dark_gray(), name.into()].to_vec(),
                        (false, true) => ["‚ñ™".dark_gray()].to_vec(),
                        (false, false) => ["‚ñ´".dark_gray()].to_vec(),
                    }))
                }
            }
        }
    }
}

impl<'a> StatefulWidget for Explorer<'a> {
    type State = ExplorerState<'a>;

    fn render(self, area: Rect, buf: &mut Buffer, state: &mut Self::State) {
        let block = Block::bordered()
            .border_type(if state.active {
                BorderType::Thick
            } else {
                BorderType::Rounded
            })
            .title_style(Style::default().italic().bold());

        let Rect { height, .. } = block.inner(area);
        state.update_offset_mut(height.into());

        let sort_symbol = match state.sort {
            Sort::Asc => SORT_SYMBOL_ASC,
            Sort::Desc => SORT_SYMBOL_DESC,
        };

        let items: Vec<ListItem> = state
            .flat_items
            .iter()
            .map(Explorer::list_item(state.selected_path(), state.is_open()))
            .collect();

        if state.is_open() {
            List::new(items)
                .block(
                    block
                        .title(format!(
                            "{} {} ",
                            if state.visibility == Visibility::FullWidth {
                                " ‚üπ "
                            } else {
                                ""
                            },
                            state.title
                        ))
                        .title(
                            Line::from(vec![" ".into(), sort_symbol.into(), " ‚óÄ ".into()])
                                .alignment(Alignment::Right),
                        ),
                )
                .highlight_style(Style::new().reversed().dark_gray())
                .highlight_symbol(" ")
                .render(area, buf, &mut state.list_state);
        } else {
            let layout = Layout::horizontal([Constraint::Length(5)]).split(area);

            List::new(items)
                .block(
                    block
                        .title(" ‚ñ∂ ")
                        .borders(Borders::LEFT | Borders::TOP | Borders::BOTTOM),
                )
                .highlight_style(Style::new().reversed().dark_gray())
                .highlight_symbol(" ")
                .render(layout[0], buf, &mut state.list_state);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use basalt_core::obsidian::VaultEntry;
    use insta::assert_snapshot;
    use ratatui::{backend::TestBackend, Terminal};

    #[test]
    fn test_render_entries() {
        let tests = [
            [].to_vec(),
            [
                VaultEntry::File(Note {
                    name: "Test".into(),
                    path: "test.md".into(),
                }),
                VaultEntry::File(Note {
                    name: "Andesite".into(),
                    path: "andesite.md".into(),
                }),
            ]
            .to_vec(),
            [VaultEntry::Directory {
                name: "TestDir".into(),
                path: "test_dir".into(),
                entries: vec![],
            }]
            .to_vec(),
            [VaultEntry::Directory {
                name: "TestDir".into(),
                path: "test_dir".into(),
                entries: vec![
                    VaultEntry::File(Note {
                        name: "Andesite".into(),
                        path: "test_dir/andesite.md".into(),
                    }),
                    VaultEntry::Directory {
                        name: "Notes".into(),
                        path: "test_dir/notes".into(),
                        entries: vec![VaultEntry::File(Note {
                            name: "Pathing".into(),
                            path: "test_dir/notes/pathing.md".into(),
                        })],
                    },
                    VaultEntry::Directory {
                        name: "Amber Specs".into(),
                        path: "test_dir/amber_specs".into(),
                        entries: vec![VaultEntry::File(Note {
                            name: "Spec_01".into(),
                            path: "test_dir/amber_specs/spec_01.md".into(),
                        })],
                    },
                ],
            }]
            .to_vec(),
        ];

        let mut terminal = Terminal::new(TestBackend::new(30, 10)).unwrap();

        tests.into_iter().for_each(|items| {
            _ = terminal.clear();
            let mut state = ExplorerState::new("Test", items);
            state.select();
            state.sort();

            terminal
                .draw(|frame| {
                    Explorer::default().render(frame.area(), frame.buffer_mut(), &mut state)
                })
                .unwrap();
            assert_snapshot!(terminal.backend());
        });
    }
}
