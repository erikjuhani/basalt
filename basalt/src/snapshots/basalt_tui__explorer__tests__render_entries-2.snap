---
source: basalt/src/explorer.rs
expression: terminal.backend()
---
"┏ Test ━━━━━━━━━━━━━━━━ ↓𝌆 ◀ ┓"
"┃   Test                     ┃"
"┃ ◆ Andesite                 ┃"
"┃                            ┃"
"┃                            ┃"
"┃                            ┃"
"┃                            ┃"
"┃                            ┃"
"┃                            ┃"
"┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛"
